File: packages/kits/GeniusLocusKit/Sources/GeniusLocusKit/Verbs/SensitivityAuditVerbs.swift
// SensitivityAuditVerbs.swift
//
// ADR-025 sensitivity unlock, §4 (Audit): "Every grant, every denial,
// every manual revocation (`mootx01 lock`), and every read served under an
// active grant is written to the UnifiedAuditLog with tier, grant id, and
// timestamps."
//
// Mirrors the FUP-C / GLK-03 grant-lifecycle audit seam in VerbSurface.swift
// (`issueGrant`/`revokeGrant`/`appendGrantAuditEntry`) — same
// non-drawer-scoped audit entry shape (a synthetic id in `rowID`, a
// descriptive token in `fieldPath`), same HLC derivation from `now`.
// Entries are durably appended through `storage.auditLog` (LocusKit's
// audit machinery) and are visible via `auditLog(for:)`, which reads
// back through `AuditBridge`'s synthetic-verb path (AuditBridge.swift).
// The four public methods below are `async throws` and await that
// append directly, so a failed write surfaces to the caller instead of
// being dropped (AUDIT-TRAIL-RESTORE-GRANTS, 2026-07-09).
// Applies sensitivity-unlock's four NEW verbs
// (`sensitivityGrantIssued`/`Denied`/`Revoked`/`sensitivityReadUnderGrant`)
// instead of reusing the federation-reserved `grantIssued`/`grantRevoked`
// (deliberately different verbs — see `UnifiedAuditLog.swift`'s enum doc
// comment).
//
// Uses `AdjectiveSensitivity` (LocusKit) for the tier rather than a new
// GeniusLocusKit- or AriaMcpKit-level type: `.restricted`/`.secret` already
// exist there, and GeniusLocusKit must not depend on AriaMcpKit (which sits
// ABOVE it in the kit topology) for a tier vocabulary this kit can express
// with an existing LocusKit type.
//
// Public (unlike the private `appendGrantAuditEntry`) because the caller
// is AriaMcpKit's `SensitivityGrantLedger`/`ToolDispatcher`, not internal
// GeniusLocusKit logic — this IS the public seam ARIA calls into. Four
// narrow, purpose-built methods (mirroring `issueGrant`/`revokeGrant`)
// rather than one generic "append any entry" method, so the audit log's
// integrity guarantees are not weakened by an unconstrained append surface.
//
// No expiry verb: expiry is passive per ADR-025 §4 — the issued record's
// `afterValue` carries its own expiry timestamp (epoch-ms, `.integer`), so
// expiry is derivable from the log without a dedicated expiry-time writer.

import Foundation
import LocusKit
import SubstrateTypes

extension GeniusLocusKit {

    /// Record that a sensitivity-unlock grant was approved and is now
    /// live. `grantID` is a fresh identifier the caller mints per grant
    /// (distinct from any drawer id — there is no natural "row" for a
    /// grant event, so a synthetic id fills the `rowID` slot, exactly as
    /// `issueGrant` does for federation grants). `expiresAt` is stored in
    /// `afterValue` as epoch-milliseconds so an expiry boundary is
    /// derivable from the log alone.
    ///
    /// - Parameters:
    ///   - handle: the estate the grant applies to. Must be open in this kit.
    ///   - tier: `.restricted` or `.secret` — any other case is a caller
    ///     error but is still recorded verbatim (this method does not
    ///     validate the tier; the ledger that calls it only ever passes
    ///     one of the two).
    ///   - grantID: a fresh, caller-minted identifier for this specific
    ///     grant instance — reused in a later `recordSensitivityGrantRevoked`
    ///     call for the SAME grant so the two entries correlate by `rowID`.
    ///   - expiresAt: the grant's computed expiry instant.
    ///   - now: issue instant: deterministic, supplied by the caller.
    /// - Throws: `GeniusLocusKitError.estateNotOpen` for a stale handle.
    public func recordSensitivityGrantIssued(
        _ handle: EstateHandle,
        tier: AdjectiveSensitivity,
        grantID: UUID,
        expiresAt: Date,
        now: Date
    ) async throws {
        _ = try estate(for: handle)
        try await appendSensitivityAuditEntry(
            verb: .sensitivityGrantIssued,
            rowID: grantID,
            fieldPath: tier.auditToken,
            before: .null,
            after: .integer(Self.epochMilliseconds(expiresAt)),
            handle: handle,
            now: now
        )
    }

    /// Record that a sensitivity-unlock grant request was DENIED (a failed
    /// `UnlockAuthority` evaluation — wrong password, cancelled biometric
    /// prompt, etc.). There is no persisted grant to correlate, so
    /// `grantID` is a fresh id minted for the denial event itself, purely
    /// so the entry has a stable, unique identity.
    public func recordSensitivityGrantDenied(
        _ handle: EstateHandle,
        tier: AdjectiveSensitivity,
        now: Date
    ) async throws {
        _ = try estate(for: handle)
        try await appendSensitivityAuditEntry(
            verb: .sensitivityGrantDenied,
            rowID: UUID(),
            fieldPath: tier.auditToken,
            before: .null,
            after: .null,
            handle: handle,
            now: now
        )
    }

    /// Record a manual revocation (`mootx01 lock`) of a live grant.
    /// `grantID` should be the SAME id passed to the
    /// `recordSensitivityGrantIssued` call this revokes, so the two
    /// entries correlate by `rowID` — mirrors `revokeGrant`'s treatment of
    /// federation grants exactly.
    public func recordSensitivityGrantRevoked(
        _ handle: EstateHandle,
        tier: AdjectiveSensitivity,
        grantID: UUID,
        now: Date
    ) async throws {
        _ = try estate(for: handle)
        try await appendSensitivityAuditEntry(
            verb: .sensitivityGrantRevoked,
            rowID: grantID,
            fieldPath: tier.auditToken,
            before: .integer(1),
            after: .null,
            handle: handle,
            now: now
        )
    }

    /// Record that a specific drawer was read ONLY because a live
    /// sensitivity grant admitted it past the default ceiling. `rowID` is
    /// the DRAWER's own id here (not a grant id) — a read-under-grant
    /// entry is genuinely about a specific row, unlike the three verbs
    /// above.
    ///
    /// - Parameter drawerID: the drawer's id, as a string (matching every
    ///   other ARIA-boundary row-id representation — MCP tool args and
    ///   results carry drawer ids as strings, not `UUID`, so the caller
    ///   passes the string it already has rather than re-parsing it).
    public func recordSensitivityReadUnderGrant(
        _ handle: EstateHandle,
        tier: AdjectiveSensitivity,
        drawerID: String,
        now: Date
    ) async throws {
        _ = try estate(for: handle)
        guard let rowUUID = UUID(uuidString: drawerID) else {
            // Malformed drawer id: should not happen (every drawer id in
            // this codebase is a UUID string), but audit recording must
            // never crash a read path — silently skip rather than throw,
            // matching the "audit is best-effort observability, not a
            // gate" posture the rest of this file's callers rely on.
            return
        }
        try await appendSensitivityAuditEntry(
            verb: .sensitivityReadUnderGrant,
            rowID: rowUUID,
            fieldPath: tier.auditToken,
            before: .null,
            after: .null,
            handle: handle,
            now: now
        )
    }

    /// Shared append helper for the four methods above.
    ///
    /// Write a sensitivity audit entry to the durable audit log (#10).
    /// These are synthetic events (not drawer mutations) — `before` /
    /// `after` are encoded into the event's bitmap slots via
    /// `SyntheticAuditValueCodec` (AuditBridge.swift) rather than real
    /// per-column bitmap state, and `AuditBridge`'s synthetic-verb path
    /// decodes them back on read. The `reason` field carries the
    /// fieldPath (tier token) so the entry is self-describing.
    ///
    /// Awaited, not fire-and-forget: this is a security-relevant write
    /// path (ADR-025 §4 requires every grant, denial, revocation, and
    /// read-under-grant to land in the audit log), so a failed durable
    /// append must surface to the caller as a thrown error rather than
    /// being silently dropped by an unstructured `Task { try? … }`. The
    /// four public methods above are `async throws` so the error
    /// propagates with no further signature change. Callers that treat
    /// audit recording as best-effort (e.g. AriaMcpKit's read-path
    /// callers) still get that posture explicitly, via `try?` at their
    /// own call site — the throw is real, but swallowing it is the
    /// caller's choice, not this method's.
    private func appendSensitivityAuditEntry(
        verb: UnifiedAuditVerb,
        rowID: UUID,
        fieldPath: String,
        before: UnifiedAuditValue,
        after: UnifiedAuditValue,
        handle: EstateHandle,
        now: Date
    ) async throws {
        guard let storage = storages[handle] else { return }
        let nowMs = Int64(now.timeIntervalSince1970 * 1000)
        // Synthetic HLC: physical time from `now`, logical 0, node 0.
        // These entries are not order-critical relative to drawer mutations
        // (they don't feed bitmap fold), so a standalone HLC is acceptable.
        let hlc = HLC(physicalTime: nowMs, logicalCount: 0, nodeID: 0)
        let (beforeKind, beforePayload) = SyntheticAuditValueCodec.encode(before)
        let (afterKind, afterPayload) = SyntheticAuditValueCodec.encode(after)
        let event = AuditEvent(
            estateUuid: handle.estateUUID,
            rowId: rowID,
            hlc: hlc,
            verb: verb.rawValue,
            beforeBitmaps: (adjective: beforePayload, operational: 0, provenance: beforeKind),
            afterBitmaps: (adjective: afterPayload, operational: 0, provenance: afterKind),
            beforeLatticeAnchor: nil,
            afterLatticeAnchor: SubstrateTypes.LatticeAnchor(udcCode: 0, qidPointer: 0),
            actor: "sensitivity-audit",
            reason: fieldPath
        )
        try await storage.auditLog.append(event)
    }

    private static func epochMilliseconds(_ date: Date) -> Int64 {
        Int64((date.timeIntervalSince1970 * 1000).rounded())
    }
}

private extension AdjectiveSensitivity {
    /// The `fieldPath` token stamped onto a sensitivity-unlock audit
    /// entry — mirrors the federation seam's "custody-mode token in
    /// fieldPath" convention (VerbSurface.swift). Only `.restricted` and
    /// `.secret` are ever passed by the sensitivity-unlock callers; the
    /// other two cases are covered for exhaustiveness, not because a
    /// grant is ever issued for them.
    var auditToken: String {
        switch self {
        case .normal: return "normal"
        case .elevated: return "elevated"
        case .restricted: return "restricted"
        case .secret: return "secret"
        }
    }
}

File: packages/kits/GeniusLocusKit/rust/tests/sensitivity_audit_verbs.rs
// sensitivity_audit_verbs.rs — out-of-band sensitivity grants (Audit) parity coverage.
//
// Mirrors Swift `SensitivityAuditVerbsTests.swift`. Same harness shape as
// grants_parity.rs's GRT-03 coordinator integration tests.

use genius_locus_kit::EstateCoordinator;
use locus_kit::adjectives::AdjectiveSensitivity;
use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::estate_types::OwnerCredentials;
use std::sync::Arc;
use uuid::Uuid;

const NOW_MS: i64 = 1_700_000_000_000;

fn open_coord() -> (EstateCoordinator, genius_locus_kit::EstateHandle) {
    let mut coord = EstateCoordinator::new();
    let store: Arc<dyn DrawerStore> = Arc::new(InMemoryDrawerStore::new(NOW_MS / 1000, None).unwrap());
    let handle = coord.open(store, OwnerCredentials::new("sensitivity-audit-owner"), 0, 100).expect("open");
    (coord, handle)
}

#[test]
fn record_sensitivity_grant_issued_emits_expected_entry() {
    let (mut coord, h) = open_coord();
    let grant_id = Uuid::new_v4();
    let expires_at_ms = NOW_MS + 30 * 60 * 1000;
    coord
        .record_sensitivity_grant_issued(&h, AdjectiveSensitivity::Secret, grant_id, expires_at_ms, NOW_MS)
        .expect("record issued");

    let log = coord.audit_log(&h).expect("audit log");
    let entries: Vec<_> = log
        .ordered_entries()
        .into_iter()
        .filter(|e| e.verb == genius_locus_kit::audit::UnifiedAuditVerb::SensitivityGrantIssued)
        .collect();
    assert_eq!(entries.len(), 1);
    assert_eq!(entries[0].field_path, "secret");
}

#[test]
fn record_sensitivity_grant_denied_emits_expected_entry() {
    let (mut coord, h) = open_coord();
    coord
        .record_sensitivity_grant_denied(&h, AdjectiveSensitivity::Restricted, NOW_MS)
        .expect("record denied");

    let log = coord.audit_log(&h).expect("audit log");
    let entries: Vec<_> = log
        .ordered_entries()
        .into_iter()
        .filter(|e| e.verb == genius_locus_kit::audit::UnifiedAuditVerb::SensitivityGrantDenied)
        .collect();
    assert_eq!(entries.len(), 1);
    assert_eq!(entries[0].field_path, "restricted");
}

#[test]
fn record_sensitivity_grant_revoked_correlates_with_issued_by_row_id() {
    let (mut coord, h) = open_coord();
    let grant_id = Uuid::new_v4();
    coord
        .record_sensitivity_grant_issued(&h, AdjectiveSensitivity::Restricted, grant_id, NOW_MS + 3_600_000, NOW_MS)
        .unwrap();
    coord
        .record_sensitivity_grant_revoked(&h, AdjectiveSensitivity::Restricted, grant_id, NOW_MS + 60_000)
        .unwrap();

    let log = coord.audit_log(&h).expect("audit log");
    let all = log.ordered_entries();
    let issued = all.iter().find(|e| e.verb == genius_locus_kit::audit::UnifiedAuditVerb::SensitivityGrantIssued).unwrap();
    let revoked = all.iter().find(|e| e.verb == genius_locus_kit::audit::UnifiedAuditVerb::SensitivityGrantRevoked).unwrap();
    assert_eq!(issued.row_id, revoked.row_id, "revoked entry must correlate with issued entry by row_id");
}

#[test]
fn record_sensitivity_read_under_grant_uses_drawer_id_as_row_id() {
    let (mut coord, h) = open_coord();
    let drawer_id = Uuid::new_v4();
    coord
        .record_sensitivity_read_under_grant(&h, AdjectiveSensitivity::Secret, &drawer_id.to_string(), NOW_MS)
        .unwrap();

    let log = coord.audit_log(&h).expect("audit log");
    let entries: Vec<_> = log
        .ordered_entries()
        .into_iter()
        .filter(|e| e.verb == genius_locus_kit::audit::UnifiedAuditVerb::SensitivityReadUnderGrant)
        .collect();
    assert_eq!(entries.len(), 1);
    assert_eq!(entries[0].row_id, genius_locus_kit::audit::EntryUUID(drawer_id.as_u128().to_be_bytes()));
}

#[test]
fn record_sensitivity_read_under_grant_silently_skips_a_malformed_drawer_id() {
    let (mut coord, h) = open_coord();
    coord
        .record_sensitivity_read_under_grant(&h, AdjectiveSensitivity::Secret, "not-a-uuid", NOW_MS)
        .expect("must not error even for a malformed drawer id");
    let log = coord.audit_log(&h).expect("audit log");
    assert!(log
        .ordered_entries()
        .into_iter()
        .all(|e| e.verb != genius_locus_kit::audit::UnifiedAuditVerb::SensitivityReadUnderGrant));
}

// NOTE (RUST-AUDIT-DURABILITY, 2026-07-09): Rust's `UnifiedAuditVerb` now
// carries `GrantIssued`/`GrantRevoked` (the FUP-C / GLK-03 grant-lifecycle
// seam durably appends them — `EstateCoordinator::issue_grant`/
// `revoke_grant`), so there is something to accidentally reuse. The Rust
// equivalent of Swift's `sensitivityGrantVerbsNeverReuseFederationReservedVerbs`
// now lives in `audit_durability_grant_sensitivity.rs`'s
// `grant_verbs_never_collide_with_sensitivity_verbs`.
